### PR TITLE
SpinLock: Overhaul false sharing prevention

### LINKED ISSUE #100163: 32-bit builds of Godot do not work with LLVM MinGW
### Tested versions

- Latest master (aa8d9b83f66488dbb2c9c918e9016ef80f821fb4)
- all versions after 4.4dev5

### System information

Godot v4.3.rc (322e93592) - Windows 10.0.26100 - GLES3 (Compatibility) - Radeon RX 580 Series (Advanced Micro Devices, Inc.; 31.0.21921.1000) - Intel(R) Core(TM) i5-9600K CPU @ 3.70GHz (6 Threads)

### Issue description

Any attempt at building Godot for 32-bit Windows will _succeed_, but then actually trying to run will crash with this

```
================================================================
CrashHandlerException: Program crashed with signal 11
Engine version: Godot Engine v4.4.dev.custom_build (55e55b4879f7ad89086aa26f2e53882cf85077f9)
Dumping the backtrace.
-- END OF BACKTRACE --
================================================================
```

And just stop. And for some bizarre reason, it _does not_ crash if its a dev build with debug symbols? (`dev_mode="yes" dev_build="yes"`)
~~I'm still not sure if this is a Godot issue or a compiler issue.~~ This is a Godot issue, see next comment.

### Steps to reproduce

1. Use the [latest](https://github.com/mstorsjo/llvm-mingw/releases/tag/20241203) LLVM MinGW (can be older too, I tested [20231128](https://github.com/mstorsjo/llvm-mingw/releases/tag/20231128)
2. Compile for windows with `arch="x86_32"`
3. Wait for compilation and try to run the executable.
4. (extra) re-compile with `dev_mode="yes" dev_build="yes"` and see how it does not crash.

### Minimal reproduction project (MRP)

Use this link to get [the builds](https://drive.google.com/drive/folders/1JScMvI4kSS9Bes-v7gVCgdHIF3ObBGKa?usp=sharing) and test it yourself:

### PR DESCRIPTION
**Pull request by RandomShaper**
This changes how false sharing in `SpinLock` is prevented. Formerly, alignment attributes were used. The problem with that is objects containing `SpinLock`s could be allocated in a way that the alignment is not honored, leading to crashes.

This PR also applies the new approach to the Apple implementation.

This approach could be refactored into a more general one that we may want to use in other sync primitives. However, the point is now to fix the only case we have at the moment.

Fixes crashes stated in https://github.com/godotengine/godot/pull/85167#issuecomment-2468695017.

*Bugsquad edit:*
- Fixes #100163.

### CODE REVIEW

**Comment by RandomShaper** [comment]
@yretenai, following up here. Could you please test this one?

File: core/os/spin_lock.h [hunk 1]
```diff
@@ -54,7 +63,7 @@ class SpinLock {
 	}
 };
 
-#else
+#else __APPLE__
```
Comment (akien-mga) [inline core/os/spin_lock.h:66]:
```suggestion
#else // __APPLE__
```

**Comment by RandomShaper** [comment]
1000 force-pushes later... 😅

**Comment by yretenai** [comment]
AVX2 and AVX512 builds are fixed with this patchset.

**Comment by RandomShaper** [comment]
@Calinou, could you please give this a go?

**Comment by RandomShaper** [comment]
Tested myself. Performance stays compared to prior to this PR, noise aside.

**Review by akien-mga — APPROVED** [review]

**Review by Calinou — APPROVED** [review]
Code looks good to me.

**Comment by Repiteo** [comment]
Thanks!

**Comment by RandomShaper** [comment]
Still, for the records, we would really like to be able to rely on alignment and base this on it. @hpvb is looking into making Godot allocation aware of alignment.

**Comment by hpvb** [comment]
@RandomShaper I'll make a pr for memory.cpp/h asap. It should be a relatively small fix.

**Pull request merged**